### ISSUE TITLE
New connections occasionally cause infinite loop in HttpHandler->Pull

### ISSUE THREAD

**Issue by ClearwaterUK** [sirn-se/websocket-php#83]
**Describe the bug**

New connections that connect to a server that is being managed by the websocket-php library have a small intermittent chance to cause the server to hang on an infinite loop, causing 100% cpu usage, and blocking server operations (including handling new and existing connections) until the server is manually restarted, or a timeout handler is implemented and triggered.
This issue is only triggered on occasion by processing new connections; connections that have already established do not trigger this problem.

While I have yet to analyse the stream data that is being recieved and parsed, I have been able to trace the source of the infinite loop to the pull() method from the HttpHandler class in the library: https://github.com/sirn-se/websocket-php/blob/5488e59e7e68d9e129a112fc7ede6e72a2c83b01/src/Http/HttpHandler.php#L48

![image](https://github.com/user-attachments/assets/d3de5b56-7236-471a-84aa-139b49afd811)

It is possible that the exit condition for the while loop is not being matched on data read from the stream with certain specific connections, causing an infinite loop and blocking further execution. 
It is plausible that non-HTTP/1.1 handshakes may be causing this but I will need to investigate further on my side to see what kind of data being sent from the remote stream is being processed.

**How to reproduce**

- Start a server with the standard middleware (CloseHandler & PingResponder)
- In HttpHandler->Pull, add some kind of echo or logging call in the do-while loop
- Have a client using any kind of Websocket library or implementation (regardless of language used) make a Handshake to the running server.
- Cross fingers and hope the connection causes the server to get stuck in the aforementioned infinite loop, causing log output spam from the previous step. If not, try again later with a different client from another location.

This issue may sometimes occur with remote connections, but I have been unable to find any one specific connection from a remote client that can trigger this, it can be a roll of the dice, but when it happens, it is clearly noticable.

**Logs**
![image](https://github.com/user-attachments/assets/20132463-c115-4f40-bc6a-d3e3faafdf1f)

**Versions**
* Version of this library: 3.2.0
* PHP version: 8.1.2 (with XDebug 3.1.2)

**Comment by ClearwaterUK** [comment]
Small elaboration: I've also confirmed this happens regardless of OS. Tested on Windows 10, Windows 11 and Ubuntu.

**Comment by sirn-se** [comment]
Hi @ClearwaterUK, thanks for reporting issue.

I haven't been able to reproduce this issue myself but I can see how this error can occur if server do not respond as expected.

If you are able to provide info on what the function _does_ read, each step in loop, it would be very helpful.
You can also print the return of `$this->stream->getMetadata()` if `$buffer` is empty after read, it might indicate some issue with the stream.

Something like this line added;
```php
        $data = '';
        do {
            $buffer = $this->stream->readLine(1024);
            echo json_encode(['buffer' => $buffer, 'meta' => $this->stream->getMetadata()]) . "\n";
            $data .= $buffer;
        } while (substr_count($data, "\r\n\r\n") == 0);
```

My idea is that if we can find a way to identify that something has gone wrong when reading the response, a `ReconnectException` should be thrown. This would force the Client to restart connect and handshake procedure.

**Comment by ClearwaterUK** [comment]
Hi again, thanks for your reply!

I've gone ahead and modified a few lines in the do-while loop to get some debug output printed as suggested, as well as adding a potential fix that may or may not help:

```php       
        $data = '';
        $recievedBytes = 0;
        do {
            $buffer = $this->stream->readLine(1024);
            $receivedBytes = strlen($buffer);
            echo ("Received " . $receivedBytes . " bytes\n");
            echo json_encode(['buffer' => $buffer, 'meta' => $this->stream->getMetadata()]) . "\n";
            $data .= $buffer;
        }  while ((substr_count($data, "\r\n\r\n") == 0) || $receivedBytes != 0);
        echo($data . "\n\n");
```

I'll keep my eye on my server today and see if a freeze happens again at all, and if it happens, what it's currently displaying.

**Comment by ClearwaterUK** [comment]
Update: I caught a server freeze, and the log output spams the following line over and over in the do-while loop shown above:

```{"buffer":null,"meta":{"timed_out":false,"blocked":true,"eof":true,"stream_type":"tcp_socket\/ssl","mode":"r+","unread_bytes":0,"seekable":false}}```

**Comment by ClearwaterUK** [comment]
I've made a couple of changes to the do-while loop on my end, and I'm happy to report that I've now been able to go more than 36 hours of uptime (without counting scheduled restarts every 12 hours) without any server freezes since.

From the logging I've added, I've noticed that there are some malformed HTTP requests that have been coming in: 

```
{"buffer":"\\u0001\\u0000\\b\\u0000\\u0003\\u0000\\u0000\\u0000","meta":{"timed_out":false,"blocked":true,"eof":true,"stream_type":"tcp_socket\\/ssl","mode":"r+","unread_bytes":0,"seekable":false}}\n\x03\x00\x00/*\xe0\x00\x00\x00\x00\x00Cookie: mstshash=Administr\r\n\x01\x00\x08\x00\x03\x00\x00\x00\n\n\
```

These appear to be RDP scan attempts from bad actors (See here for more info https://github.com/olipo186/Git-Auto-Deploy/issues/221 ), and may have been what was causing infinite loops to happen as the library was checking for its original exit condition in the recieved data where it had none.

**Comment by sirn-se** [comment]
Good find, it certainly appears to be the RDP scan you mention.

What changes did you do to the read loop?

**Comment by ClearwaterUK** [comment]
These are the current changes I've made to the do-while loop:
```php
        $data = '';
        $unreadBytes = 0;
        $streamMetadata = "";
        do {
            $buffer = $this->stream->readLine(1024);
            $streamMetadata = json_decode(json_encode($this->stream->getMetadata()));
            $unreadBytes = $streamMetadata->{'unread_bytes'};
            $data .= $buffer;
        } while ($unreadBytes != 0);
```
The main change is decoding the stream metadata into an object, and verifying that no new bytes have come in, which signifies the end of the data stream.
There likely is a more robust way of implementing this fix (I hate that I have to encode then re-decode the stream metadata), but I can at least confirm these changes work.

If you're okay with this, I can shoot a PR your way.

**Comment by sirn-se** [comment]
What worries me a bit is that PHP documentations for `unread_bytes` says "Note: You shouldn't use this value in a script.".
https://www.php.net/manual/en/function.stream-get-meta-data.php

No explanation why, though. Going to check in on this a bit.

You can provide key when getting metadata, like `$this->stream->getMetadata('unread_bytes')`. Then the json back-and-forth is unnecessary.

**Comment by ClearwaterUK** [comment]
> What worries me a bit is that PHP documentations for unread_bytes says "Note: You shouldn't use this value in a script.".

Good catch there, although I have noticed `eof` may be able to be used in its place instead.

**Comment by sirn-se** [comment]
Tried a bit different approach, does catch the malformed HTTP exploit you experienced: https://github.com/sirn-se/websocket-php/pull/89

Basically, it validates the first line and exits if not valid http, without attempting to read more of request.
It also fails if expected data in request is not readable.

I'm also considering moving handshake to the stream-select routine. This way, reading HTTP request will not be blocking other connections. But larger change so will take that one later.

**Comment by sirn-se** [comment]
`v3.2.1` just released, resolves this issue.
